feat: grpc call for compact object store chunks

NGA-TRAN opened the pull request:
Closes #3124
Another piece of #3089 

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

NGA-TRAN commented on generated_types/protos/influxdata/iox/management/v1/service.proto:
```diff
@@ -90,6 +90,11 @@ service ManagementService {
+  // Compact given object store chunks 
+  //
+  // Errors if the chunks are not compacted yet and not contiguous
+  rpc CompactObjectStoreChunks(CompactObjectStoreChunksRequest) returns (CompactObjectStoreChunksResponse);
```
Now I have this ready, after the fix in https://github.com/influxdata/influxdb_iox/pull/3289 is merged, I will add test for this client as well as end-to-end tests in next PR.

alamb commented on generated_types/protos/influxdata/iox/management/v1/service.proto:
```diff
@@ -472,3 +477,22 @@ message DropPartitionRequest {
+}
+
+message CompactObjectStoreChunksResponse {
+}
```
I wonder if the request should return a JobId (so that the compaction can happen as an async task rather than a synchronous one). I can imagine some compaction operations might take longer than the 30 second gRPC timeout

Something like this (following the model of `RebuildCatalog`):

```suggestion
  // The operation that tracks the work for compacting the chunks
  google.longrunning.Operation operation = 1;
}
```

https://github.com/influxdata/influxdb_iox/blob/main/generated_types/protos/influxdata/iox/management/v1/service.proto#L429-L432

NGA-TRAN replied:
Great suggestion and link. Working on this

alamb commented on influxdb_iox/src/commands/database/partition.rs:
```diff
@@ -67,6 +67,25 @@ struct Persist {
 
+/// Compact Object Store Chunks
+///
+/// Errors if the chunks are not yet compacted and not contiguous. If successful it returns the
+/// compacted chunk or None if no data after compacted due to rows eliminated from deletes and deduplication
```
i don't think this comment is 100% accurate anymore -- the command doesn't seem to return the compacted chunk (it seems to print `"Ok"` on success).

NGA-TRAN replied:
Good catch

alamb commented on influxdb_iox/src/commands/database/partition.rs:
```diff
@@ -152,6 +171,12 @@ enum Command {
+    /// Compact Object Store Chunks
+    ///
+    /// Errors if the chunks are not yet compacted and not contiguous. If successful it returns the
+    /// compacted chunk or None if no data after compacted due to rows eliminated from deletes and deduplication
```
likewise here the reference to returning the compacted chunk I think is erronious

alamb commented on influxdb_iox/src/influxdb_ioxd/server_type/database/rpc/management.rs:
```diff
@@ -534,6 +534,38 @@ impl management_service_server::ManagementService for ManagementService {
+            chunk_id_ids.push(chunk_id);
+        }
+
+        db.compact_object_store_chunks(&table_name, &partition_key, chunk_id_ids)
```
As I mentioned above, since I think this operation can take quite a while, doing it as an asyncrhonous Job and returning a job ID might make more sense here rather than waiting for the job to finish.

alamb commented on influxdb_iox_client/src/client/management.rs:
```diff
@@ -503,4 +503,34 @@ impl Client {
+        let db_name = db_name.into();
+        let partition_key = partition_key.into();
+        let table_name = table_name.into();
+        // let chunk_ids = chunk_ids
```
Left over?

alamb reviewed (approved):
I do think making `compact_chunks` an async operation (by returning a JobId) is / will be important. However, I also think it could be done as a follow on PR as well.

NGA-TRAN commented:
@alamb O have addressed all your comments. Can you have a quick look to see if they work the way you suggested? I  adding tests to verify the job returned messages before merging

alamb commented on generated_types/protos/influxdata/iox/management/v1/service.proto:
```diff
@@ -472,3 +477,24 @@ message DropPartitionRequest {
+
+message CompactObjectStoreChunksResponse {
+  // The operation that tracks the work for compacting object store chunks
+  google.longrunning.Operation operation = 1;
```
👍

alamb reviewed (approved):
Looks good. Thanks @NGA-TRAN
